### ISSUE TITLE
To implement ToSocketAddrs for a custom type the ToSocketAddrsFuture should be pub use

### ISSUE THREAD

**Issue by bandheight** [async-rs/async-std#539]

**Comment by yoshuawuyts** [comment]
Ref https://docs.rs/async-std/1.0.1/async_std/net/trait.ToSocketAddrs.html. From a glance it seems this is indeed accurate. Since the result type from `ToSocketAddrs` is opaque, we could probably change it to be similar to [`FromStream::from_stream`](https://docs.rs/async-std/1.0.1/async_std/stream/trait.FromStream.html#tymethod.from_stream) without breaking anything.

@bandheight out of curiosity: what's your use case for implementing `ToSocketAddrs`?

**Comment by stjepang** [comment]
I'd be okay with changing the return type of `to_socket_addrs` to `Pin<Box<dyn Future>>` here. Performance impact of the change would be effectively none.

**Comment by bandheight** [comment]
@yoshuawuyts I implemented an `enum` type to represent multiple address types.

```rust
pub enum Address {
    /// IP Address
    SocketAddr(SocketAddr),
    /// Domain name address, eg. example.com:8080
    Domain(String, u16),
}
```
If this type implement the trait `ToSocketAddrs`,  it will be passed directly to `TcpListener::bind` and other similar functions.

**Comment by yoshuawuyts** [comment]
@bandheight does the following work for you?

```rust
async fn bind_socket(addr: Address) -> io::Result<TcpListener> {
    match addr {
        SocketAddr(addr) => TcpListener::bind(addr).await?,
        SocketAddr((addr, port) => TcpListener::bind((&*addr, port)).await?,
    }
}
```

If we don't already support `ToSocketAddr` for `(String, u16)` then we should add that!

**Comment by bandheight** [comment]
@yoshuawuyts This code is worked. 
And it is also my current solution, but I want to make my code look more neat and elegant.
So implement the trait `ToSocketAddrs` is probably the best solution.

```rust
pub struct DomainName(pub String, pub u16);

...
// implement Deserialize for DomainName
...

#[derive(Deserialize)]
pub enum Address {
    /// IP Address
    SocketAddr(SocketAddr),
    /// Domain name address, eg. example.com:8080
    Domain(DomainName),
}

#[derive(Deserialize)]
pub struct Config {
    pub addr: Address,
}

...
let config: Config = <deserialize from config file>;
...

// This is neat and elegant.
let listener = TcpListener::bind(config.addr);
```

I vote for it that support `ToSocketAddr` for `(String, u16)`.

**Comment by dizda** [comment]
agreed with @bandheight, that'd be cleaner to implement ToSocketAddrs.

use case: 
- https://github.com/dizda/fast-socks5/blob/429d491364ca515f1d3d4ec0fefadb6c35e82bcf/src/util/target_addr.rs#L76
- https://github.com/dizda/fast-socks5/blob/429d491364ca515f1d3d4ec0fefadb6c35e82bcf/src/server.rs#L476

👍

**Comment by dizda** [comment]
Shadowsocks does it for a similar purpose, but with std::net::ToSocketAddrs.

cf. https://github.com/shadowsocks/shadowsocks-rust/blob/f4feba9fc1c8a968d90a592c5f4e86ffc6dd322b/src/relay/socks5.rs#L332
